FHEM/10_BindingsIo.pm

```diff
@@ -18,64 +17,77 @@ use CoProcess;
 use JSON;
 use Time::HiRes qw(time);
 
-sub Log($$);
-sub Log3($$$);
+our $init_done;
+our $modules;
+
 
 my $USE_DEVIO_DECODEWS = 0;
 my $timeouts = 0;
 
-sub
-BindingsIo_Initialize($)
-{
+sub BindingsIo_Initialize {
   my ($hash) = @_;
 
   $hash->{parseParams} = 1;
 
-  $hash->{DefFn}    = 'BindingsIo_Define';
-  $hash->{UndefFn}  = 'BindingsIo_Undefine';
-  $hash->{GetFn}    = 'BindingsIo_Get';
-  $hash->{SetFn}    = 'BindingsIo_Set';
-  $hash->{AttrFn}   = 'BindingsIo_Attr';
+  $hash->{DefFn}    = \&BindingsIo_Define;
+  $hash->{UndefFn}  = \&BindingsIo_Undefine;
+  $hash->{GetFn}    = \&BindingsIo_Get;
+  $hash->{SetFn}    = \&BindingsIo_Set;
+  $hash->{AttrFn}   = \&BindingsIo_Attr;
   $hash->{AttrList} = $readingFnAttributes;
-  $hash->{NotifyFn}   = 'BindingsIo_Notify';
+  $hash->{NotifyFn}   = \&BindingsIo_Notify;
 
-  $hash->{ReadFn}   = 'BindingsIo_Read';
-  $hash->{ReadyFn}  = 'BindingsIo_Ready';
-  $hash->{WriteFn}  = 'BindingsIo_Write';
+  $hash->{ReadFn}   = \&BindingsIo_Read;
+  $hash->{ReadyFn}  = \&BindingsIo_Ready;
+  $hash->{WriteFn}  = \&BindingsIo_Write;
 
   $hash->{Clients} = "PythonModule:fhempy"; # NodeModule
 
-  return undef;
+  return ;
 }
 
-sub
-BindingsIo_Define($$$)
-{
-  my ($hash, $a, $h) = @_;
+sub BindingsIo_Define {
+  my ( $hash, $a ) = @_;
+ 
   my $name = $hash->{NAME};
+  Log3 $name, 3, q[BindingsIo v1.0.1];
 
-  Log3 $hash, 3, "BindingsIo v1.0.0";
-
-  my $bindingType = @$a[2];
+  my $bindingType = $a->[3] // $a->[2];
 
   $hash->{args} = $a;
-  $hash->{argsh} = $h;
+  # $hash->{argsh} = $h; // Todo: Delete other code references
 
   my $port = 0;
   my $localServer = 1;
-  if ($bindingType eq "Python" or $bindingType eq "fhempy") {
-    $hash->{DeviceName} = "ws:127.0.0.1:15733";
-    $hash->{IP} = "127.0.0.1";
-    $hash->{PORT} = "15733";
-    $hash->{localBinding} = 1;
-  } else {
-    $hash->{DeviceName} = "ws:".@$a[2];
-    $hash->{IP} = substr($hash->{DeviceName}, 0, index($hash->{DeviceName}, ":"));
-    $hash->{PORT} = substr($hash->{DeviceName}, index($hash->{DeviceName}, ":")+1);
-    $bindingType = @$a[3];
+
+  if ($#{$a} == 3) {
+    $hash->{DeviceName} = "ws:".$a->[2];
     $localServer = 0;
     $hash->{localBinding} = 0;
+  } elsif ($#{$a} == 2) {
+    $hash->{DeviceName} = "ws:localhost:15733";
+    $hash->{localBinding} = 1;
+  } else {  
+    return q[to few parameters given for define];
   }
+
+  ( undef, $hash->{IP},$hash->{PORT} ) = split ':', $hash->{DeviceName};
+
+
+  if ( $hash->{IP} !~ m/^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/ )
+  {
+     my $ip= gethostbyname($hash->{IP});
+     if (defined $ip ) { $hash->{IP} =  inet_ntoa($ip) };
+     Log3 $name, 3, qq[found IP $hash->{IP}];
+  }
```

Comment:
Finde ich super, dass damit dann auch DNS funktioniert! 👍

Reply:
Gut, an dieser Stelle möchte ich allerdings anmerken,  dass das automatische Anlegen von Definitionen über autocreate realisiert werden sollte, damit User die Wahloption haben.

Ich habe das Szenario vermutlich auch nicht verstanden, wann es sinnvoll ist, dass der Python Prozess ein define auslöst. In meinen Fällen muss immer erst etwas in FHEM definiert werden, damit der Python Prozese etwas an FHEM übergeben kann.

Reply:
Es soll dadurch immer möglich sein, dass man einen Peer Prozess startet und in FHEM wird automatisch der fhempy_peer angelegt welcher sofort verwendet werden kann.

Der Hintergrund zu Peers:
Sie ermöglichen beliebiges hin- und herschieben von Devices zwischen unterschiedlichen Peers. Use Cases: Bluetooth Reichweite, Ressourcenauslastung verteilen, etc.. Sogar die Verlagerung in die Cloud wäre möglich.

Jede fhempy Instanz lauscht im Netzwerk via Zeroconf nach _fhempy._http._tcp.local., sobald das gefunden wird, wird der Peer in FHEM angelegt.

Die Anlage nur wenn autocreate aktiv ist, macht wahrscheinlich Sinn.

Reply:
Die Anlage über autocreate funktioniert über ein spezielles Event und nicht über den Define Befehl :) 

Das müsste aber von dem fhempy Server übermittelt werden. Kann der Events triggern?

Reply:
Ja, aus fhempy kann man auch Events generieren. Das mit autocreate hatte ich mir noch nie angesehen, weil es sonst noch eine Einstellung mehr wäre auf die man achten muss. Generell ist autocreate aber wahrscheinlich eh bei vielen Usern  aktiv.
Autocreate kann man sich ja noch separat überlegen ob es Sinn macht. Auch die Anlage von tuya Devices passiert zur Zeit automatisch und wird nicht über autocreate gemacht. Aus meiner Sicht müsste FHEM hier eine Möglichkeit schaffen, dass Devices angezeigt werden, aber nicht automatisch angelegt, die User sollend dann entscheiden ob es angelegt werden soll.

Reply:
Autocreate kann ja konfiguriert werden, was es anlegen darf oder nicht.

Anzeige ist vielleicht über das Logfile einfach möglich oder es wird eine eigene Anzeige im BindingsIo Modul erzeugt. Das mit der Anzeige ist halt ein generelles FHEM Problem. Das müsste ja eigentlich durch autocreate zentral erfolgen.